osutil/export_test.go

```diff
@@ -194,12 +195,15 @@ func MockEtcFstab(text string) (restore func()) {
 
 // MockUname mocks syscall.Uname as used by MachineName and KernelVersion
 func MockUname(f func(*syscall.Utsname) error) (restore func()) {
+	/* The old code left for comparison
 	old := syscallUname
 	syscallUname = f
 
 	return func() {
 		syscallUname = old
 	}
+	*/
+	return testutil.MockAny(&syscallUname, f)
```

Comment:
my preference is for a simpler helper that does less, but doesn't come with questions about the name, the order of arguments or moving some of the errors away from compile time:

```
r = testutil.ValueRestore(&syscallUname)
syscallUname = f
return r
```

this removes the most boring part of these mock functions which is dealing with the old value. As for the going from 3 lines to 1m is not a obvious clear win because of the other issues. As I said elsewhere these helpers are common but not that common overall.

We could also make the helper take multiple pointers,  that would help with places like MockPruneInterval that deal with multiple mocking at the same time:

```
r := ValuesRestore(&pruneInterval, &pruneWait, &abortWait)
...
```

Reply:
Hi @pedronis, 
I am sincerely trying to understand. So, please don't be angry with me for not getting it yet.
I am not sure I understand why this one is better. It won't remove the complexity of generics semantics. 
All we do here is just moving the substitution line out of the mocking function. 

By the way, the variadic argument expansion is possible in case of `testutil.MockAny(...)` too.

Please, help me understand?

Reply:
Give the trade offs,  this is quite reasonable:

```
func MockPruneInterval(prunei, prunew, abortw time.Duration) (restore func()) {
        r := testutil.ValuesRestore(&pruneInterval, &pruneWait, &abortWait)
        pruneInterval = prunei
        pruneWait = prunew
        abortWait = abortw
        return r
}
```

Reply:
Thank you for the elaborate answer. I appreciate a lot.

Reply:
@pedronis, thank you for the elaborate answer. I appreciate it a lot. :pray:

Reply:
I can expand a bit more:

* in general compile time errors are preferable to test errors which again are preferable to runtime/production errors
* panics are usually best avoided, panics are ok/might be ok: 
    * not continuing with the program execution is really the preferable behavior
    * situations concerning static/programming errors that should be generally ruled out by natural usage and/or unit tests 
      where API-wise returning errors is really more brittle than a panic
    * in doubt returning an error is generally preferable
* we are a bit more flexible with the 2nd type of situations in tests, but still panics and testing framework/our state lock might mix badly sometimes, so solutions with no/less panics are preferable even in tests if the extra cost is reasonable

Reply:
@pedronis, can we have both solutions? 

I mean these two can be very useful together. 
- For case with multiple mocking your API is the best
- In cases when mocking one thing the API I suggested could be very useful too

What do you think?

Thank you,
Arseniy

Reply:
@pedronis, I am sorry, I wrote my question above before your reply got to me. I think I have my answer. Thank you. :pray: